zebra-state/src/on_disk/block_index.rs

```diff
@@ -0,0 +1,159 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap, HashMap},
+    sync::{Arc, Weak},
+};
+use zebra_chain::serialization::{ZcashDeserialize, ZcashSerialize};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+pub(super) struct BlockIndex {
+    cache: WeakCache,
+    storage: sled::Db,
+}
+
+impl Default for BlockIndex {
+    fn default() -> Self {
+        let config = sled::Config::default();
+        Self {
+            cache: Default::default(),
+            storage: config.open().unwrap(),
+        }
+    }
+}
+
+impl BlockIndex {
+    pub(super) fn insert(
+        &mut self,
+        block: impl Into<Arc<Block>>,
+    ) -> Result<BlockHeaderHash, Error> {
+        let block = block.into();
+        let hash: BlockHeaderHash = block.as_ref().into();
+        let height = block.coinbase_height().unwrap();
+
+        // make sure we can insert into the cache first
+        self.cache.insert(block.clone())?;
+
+        let by_height = self.storage.open_tree(b"by_height")?;
+        let by_hash = self.storage.open_tree(b"by_hash")?;
+
+        let mut bytes = Vec::new();
+        block.zcash_serialize(&mut bytes)?;
+
+        // TODO(jlusby): make this transactional
```

Comment:
sled has support for transactions which would let us atomically insert into both tables, but it makes it fucks with the error handling because they only implemented the `Transactional` trait for closures that return [`Result<T, ConflictableTransactionError<()>>.`](https://docs.rs/sled/0.31.0/sled/enum.ConflictableTransactionError.html), but this doesn't implement `std::error::Error`!

It doesn't seem like we really need it though, at least not initially so this should be fine for a while